src/cswinrt/strings/additions/Windows.Storage.Streams/WindowsRuntimeMarshal.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace System.Runtime.InteropServices.WindowsRuntime
+{
+    using System;
+    using System.Diagnostics;
+    using System.Diagnostics.CodeAnalysis;
+    using global::Windows.Foundation;
+    using global::Windows.Storage.Streams;
+    using WinRT;
+
+#nullable enable
+    /// <summary>
+    /// An unsafe class that provides a set of methods to access the underlying data representations of WinRT types.
+    /// </summary>
+#if EMBED || !NET // for netstandard this type conflicts with the type in the BCL so make it internal
+    internal
+#else
+    public
+#endif
+    static partial class WindowsRuntimeMarshal
+    {
+        /// <summary>
+        /// Returns a pointer to the underlying data representation of the <see cref="IBuffer"/>.
+        /// Callers are responsible for ensuring that the buffer is kept alive while the pointer is in use.
+        /// </summary>
+        /// <param name="buffer">The buffer to get the data pointer for.</param>
+        /// <param name="dataPtr">The pointer to the underlying data representation of the buffer.</param>
+        public static unsafe bool TryGetDataUnsafe(
+#if NET
+            [NotNullWhen(true)]
+#endif
+            IBuffer? buffer, out IntPtr dataPtr)
+        {
+            if (buffer == null)
+            {
+                dataPtr = IntPtr.Zero;
+                return false;
+            }
+
+            if (ComWrappersSupport.TryUnwrapObject(buffer, out var unwrapped) &&
+                unwrapped.TryAs(global::WinRT.Interop.IID.IID_IBufferByteAccess, out IntPtr ThisPtr) >= 0)
+            {
+                try
```

Comment:
We could optimize this to not need this `try/finally`:
```csharp
IntPtr __retval;
int hresult = (*(delegate* unmanaged[Stdcall]<IntPtr, IntPtr*, int>**)ThisPtr)[3](ThisPtr, &__retval);
dataPtr = __retval;

Marshal.Release(ThisPtr);

global::WinRT.ExceptionHelpers.ThrowExceptionForHR(hresult);

return true;
```

But also, why are we throwing in the first place here, given the method is named "Try"?

Reply:
Try doesn't mean we never throw. Try here means we are not throwing if the type doesn't implement IBufferByteAccess. If it does implement it but the implementation is bad we still throw.

Reply:
I don't think it's an optimization though.
`try/finally` doesn't have performance impact for fast path (no exception being thrown) and it can make sure the finally block being executed once there's an exception. You can check the codegen, unless you don't have `catch` blocks in a `try`, it shouldn't make any meaningful impact on the performance.